presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java

```diff
@@ -283,11 +283,13 @@ public HivePageSink(
     }
 
     @Override
-    public Collection<Slice> finish()
+    public CompletableFuture<Collection<Slice>> finish()
     {
         // Must be wrapped in doAs entirely
         // Implicit FileSystem initializations are possible in HiveRecordWriter#commit -> RecordWriter#close
-        return hdfsEnvironment.doAs(session.getUser(), this::doFinish);
+        // TODO: think about processing it asynchronously
```

Comment:
Remove this TODO. It's correct to do this processing inline, since it can use substantial CPU, and won't tie up unrelated threads (like Raptor backups).